### PR TITLE
Adds option to relocate #functions before #usage

### PR DESCRIPTION
**Pull request by gadenbuie**
I have a number of reference topic pages that document a family of functions within the same reference page, since they share arguments and functionality. I frequently find that, in these cases, it can be intimidating or confusing to open with the **usage** section. Instead the **Functions** section, which is a list of functions with their description (from the `#' @describeIn <parent> <Function description>` roxygen tag), generally provides a more gentle and orienting introduction to the functions in the reference page.

Rather than try to address this at the pkgdown level, this PR "fixes it in post-production" by adding a small amount of JS to rewrite the reference page if the **Functions** section is present. This is turned off by default, but can be opted in via a `template$params$relocationFunctions` option in `_pkgdown.yml`. The one non-opt-in change is that I've added a **Usage** heading above what was previously just `<pre class="usage">`.

Note: I'm open to renaming the param or anything else in here.

* Adds `content-reference-topic.html` template from pkgdown BS3 template
* Updates template to add Usage heading above usage section
* Adds JavaScript to relocate #functions before #usage and update the TOC

This is opt-in and has to be enabled with

```yaml
template:
  params:
    relocateFunctions: true
```

Here's a quick example (which may not be the best because it's rather obvious... but you can see that the **Functions** section comes very late in the page otherwise).

<details><summary>Before</summary>

image

</details>

<details><summary>After</summary>

image

</details>

### CODE REVIEW

**Comment by cderv** [comment]
Some feedback on this proprosal:

> Rather than try to address this at the pkgdown level

Do you think it is not worth seeing if it has a place in pkgdown or it is just easier to modify **quillt** directly ? 

Rgearding the How, one drawback of changing a full template file of pkgdown is that we will need to keep it up to date in case the one from pkgdown evolves (bug fix or new feature). For example, it could be the case with the switch to BS4 as they are providing new templates.

Dev **pkgdown** has now a way to customize templates without changing the all template using `in-header` pieces for example - this would mean that a `in-header-reference-topic.html` would allow to insert JS in the header of the page, or `after-body-reference-topic.html` to add it at the end of the body (this would be equivalent to the current addition)

What prevent to do that maybe is the fact you need to add a Title header for Usage section. But this could be added by JS also maybe. 

Regarding the Why and the  need for this change, I don't think we have a lot of documentation using `@describeIn` - usually we group function in same topic but don't use this - we use [the other](https://r-pkgs.org/man.html#multiple-man) `@rdname` to do that. No Functions section is created and we often ad sections to describe with more details the differences. For example, https://pkgs.rstudio.com/rmarkdown/reference/render_site.html : Would you have a Functions sections at the top in such pages ? 

I understand this could lead to have a Usage header in all reference page, between description on the page and the code block. Is that right ? This is a big change in layout. Why not ? but when it start by this block, which will be the case for most of the reference pages, we'll have this new title where is not really needed I feel. 
Adding this head dynamically by JS would allow to have in only on those pages. It would be totally opt-in.

At last, I know this is optional but it adds some maintenance in the template that it would be good to reduce. 

If we agree to add this optional feature, I would use the insertion template, and use only JS to modify the page.

**Review by cderv — COMMENTED** [review]
Some feedback on this proprosal:

> Rather than try to address this at the pkgdown level

Do you think it is not worth seeing if it has a place in pkgdown or it is just easier to modify **quillt** directly ? 

Rgearding the How, one drawback of changing a full template file of pkgdown is that we will need to keep it up to date in case the one from pkgdown evolves (bug fix or new feature). For example, it could be the case with the switch to BS4 as they are providing new templates.

Dev **pkgdown** has now a way to customize templates without changing the all template using `in-header` pieces for example - this would mean that a `in-header-reference-topic.html` would allow to insert JS in the header of the page, or `after-body-reference-topic.html` to add it at the end of the body (this would be equivalent to the current addition)

What prevent to do that maybe is the fact you need to add a Title header for Usage section. But this could be added by JS also maybe. 

Regarding the Why and the  need for this change, I don't think we have a lot of documentation using `@describeIn` - usually we group function in same topic but don't use this - we use [the other](https://r-pkgs.org/man.html#multiple-man) `@rdname` to do that. No Functions section is created and we often ad sections to describe with more details the differences. For example, https://pkgs.rstudio.com/rmarkdown/reference/render_site.html : Would you have a Functions sections at the top in such pages ? 

I understand this could lead to have a Usage header in all reference page, between description on the page and the code block. Is that right ? This is a big change in layout. Why not ? but when it start by this block, which will be the case for most of the reference pages, we'll have this new title where is not really needed I feel. 
Adding this head dynamically by JS would allow to have in only on those pages. It would be totally opt-in.

At last, I know this is optional but it adds some maintenance in the template that it would be good to reduce. 

If we agree to add this optional feature, I would use the insertion template, and use only JS to modify the page.

**Comment by gadenbuie** [comment]
> Do you think it is not worth seeing if it has a place in pkgdown or it is just easier to modify **quillt** directly ?

I think that fixing this in pkgdown is probably too large of an undertaking. The problem is that this section comes through as a generic `tag_section` so [pkgdown follows](https://github.com/r-lib/pkgdown/blob/9c95f00d2505ddd83c4722e019202370715d3a3d/R/build-reference.R#L312-L315) the placement used in the R help page.

I use `@describeIn` often because it gives you a place to specifically describe each function in the docs. I noticed now, though, thanks to your link, that a similar section is also created for [methods and generics](https://r-pkgs.org/man.html#multiple-man) but with a different heading name.

> Dev pkgdown has now a way to customize templates without changing the all template...

This would be ideal and was my first choice as well. I may have missed something, but it seems that you have to choose between a template provided by a package OR a local customizations. If it's possible, I'd be happy to keep this `gradethis` specific, but I couldn't get both `quillt` and template overrides to work together, hence this PR.

> Regarding the Why and the need for this change...we often add custom sections to describe with more details the differences.

This is a good idea, but the real problem, IMHO, is that **usage** comes too early. The usage section is most valuable to users who already have a strong idea about how the function works and want to check the function signature.

I think that its placement makes sense in the help pages in the R console, but in the package's website users are often _learning_ about a package and its functions for the first time. I think that it can be confusing to showcase the _function signature_ above the fold (in the first screen of the reference topic).

gradethis is certainly at an extreme here in that many functions are designed to work within a specific context or environment where a bit of magic is used to set the default values depending on the context. The function signature is at best confusing until you have some understanding of what the function does.

> I don't think we have a lot of documentation using `@describeIn` - usually we group function in same topic but don't use this - we use [the other](https://r-pkgs.org/man.html#multiple-man) `@rdname` to do that. ... For example, [pkgs.rstudio.com/rmarkdown/reference/render_site.html](https://pkgs.rstudio.com/rmarkdown/reference/render_site.html) : Would you have a Functions sections at the top in such pages ?

As an aside, I think that you should consider using `@describeIn`. While the `@return` (Value) section explains what each function does, `@describeIn` gives a nice clean way to generate a list with a description of each of the functions. This is especially helpful when you use `@return` to document the object being returned, in which case that section may not lend itself to cleanly describing each function. I also appreciate having a dedicated place to look to find out what _each_ function does when multiple function appear in the same page.


### Next steps

Let's start by seeing if its possible to locally override or modify a template provided by a package. If so, I'll keep this in gradethis for now, using minimal JavaScript to make the change.

Otherwise, we can adjust the approach here to touch the template as little as possible and just make the changes with JavaScript.

**Comment by gadenbuie** [comment]
Okay, I brought the after/before body template overrides into quillt by adding `layout.html` and then used `after-body-reference-topic.html` to insert the JavaScript. In this iteration, it is still opt-in via the template parameter. I also slightly refactored the JavaScript: we add **Usage** header to all (if you've opted-in) for consistency, otherwise that section doesn't get a header if you're not relocating the functions section.

**Comment by cderv** [comment]
> This would be ideal and was my first choice as well. I may have missed something, but it seems that you have to choose between a template provided by a package OR a local customizations. If it's possible, I'd be happy to keep this gradethis specific, but I couldn't get both quillt and template overrides to work together, hence this PR.

I did not mean local but provided by **quillt**. However, I think **quillt** can only provide the additional template and not have a copy of one of the template inside **pkgdown**. Here it would mean that layout.html would not be required as it is already in **pkgdown**. 

I'll look into that to confirm ! 

About the layout change to have Usage header in all reference page, this is something impactful, that would make all quillt-powered website different from others. I let @apreshill give here feedback about this new layout proposal.

**Comment by gadenbuie** [comment]
> Here it would mean that layout.html would not be required as it is already in **pkgdown**.

Oops! I tried a few configurations of things and thought this was needed, but you're right it is not. I removed `layout.html` since it's a direct copy from pkgdown.

File: inst/pkgdown/templates/after-body-reference-topic.html [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{{#yaml}}{{#relocateFunctions}}<script type="application/javascript">
```
Comment (cderv) [inline inst/pkgdown/templates/after-body-reference-topic.html:1]:
```suggestion
{{#yaml.relocateFunctions}}<script type="application/javascript">
```

File: inst/pkgdown/templates/after-body-reference-topic.html [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  var fns = document.getElementById('functions');
+  if (!fns) { return; }
+
+  fns = [fns, fns.nextElementSibling];
+
+  // Move Functions before Usage
+  fns.forEach(function(el) {
+    usageH2.parentElement.insertBefore(el, usageH2);
+  });
+})();</script>{{/relocateFunctions}}{{/yaml}}
```
Comment (cderv) [inline inst/pkgdown/templates/after-body-reference-topic.html:28]:
```suggestion
})();</script>{{/yaml.relocateFunctions}}
```

File: inst/pkgdown/templates/after-body-reference-topic.html [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{{#yaml}}{{#relocateFunctions}}<script type="application/javascript">
+(function() {
```
Comment (cderv) [inline inst/pkgdown/templates/after-body-reference-topic.html:2]: I would add a comment to easily know in the source that this comes from quillt
```suggestion
<!-- quillt post-processing to relocate Functions sections in reference pages -->
(function() {
```

**Review by cderv — APPROVED** [review]
Last review bit and it is good for me. 

* The `.` syntax comes from issue we encountered as a named variable is looked up in context stack and not just parent content, unless dot syntax is used. See #7 and associated pkgdown issue. I would now try to use it to prevent any future issue.

* About the name, why use camelCase ? We could use `snake_case` as it feels more aligned with other YAML parameters (like `asset_path` or `default_assets`) ? 
```yaml
template:
  package: quillt
  params:
    relocate_functions: true
```

We need @apreshill review and thoughts before I merge this

**Comment by gadenbuie** [comment]
Updated to use `relocate_functions` as the template parameter name, `.` for accessing `yaml.relocate_functions` in the reference topic after-body template, and to include a comment in the HTML.

**Comment by apreshill** [comment]
I think we might want to ask for broader review. I definitely see the use case, but I haven't seen many packages in tidyverse/r-lib/or our group that have a dedicated functions section. My main concern would be that our motivation is to be least surprising to users across packages, so we'd like to avoid any deviations that might violate user expectations. Also, I'm wondering about the experience of getting help from the pkgdown reference site vs. the R help page- this will mean the online docs won't mirror the help page accessed via `?pass_if` for example, right?

**Comment by gadenbuie** [comment]
> My main concern would be that our motivation is to be least surprising to users across packages, so we'd like to avoid any deviations that might violate user expectations. Also, I'm wondering about the experience of getting help from the pkgdown reference site vs. the R help page- this will mean the online docs won't mirror the help page accessed via `?pass_if` for example, right?

That's right, in the local help pages Functions appears pretty far down. AFAIK pkgdown does some re-ordering of things but not much. I'd see this being a hard sell broadly, which is why I didn't pitch it in pkgdown.

My three main thoughts are this: 

1. First, many packages simply replicate the documentation online. That's fine. But with some additional work, the Reference Index page can actually be a pretty decent road-map for the package as a whole. With a bit of description, consistent grouping, etc, it serves a much better entry than the index page a user will find in their help page.

2. Second, along those lines, the web version gives the user _many more_ opportunities to find what they're looking for. In particular, the TOC makes jumping around much easier than in the local help pages. The pkgdown site is fundamentally better for exploring and learning about the functions in a package than the local help pages.

   Here's an example from one of the package sites featured on the pkgdown front page: <http://mc-stan.org/bayesplot/reference/PPC-distributions.html>. There are 10 functions described in this help page and the opening description immediately instructs the reader to scroll down (past the _long_ usage section) to find a sensible description of those functions. 

   If your first introduction to these functions is what you see when you open that page, you're really not ready to understand several pages of the function signature. In the bayesplot example, it makes sense for the author to write a custom section explaining the function grouping, but often we can make use of the `@describeIn` tag to do this for us cleanly.

3. The problem is definitely exacerbated in gradethis, where for many functions you are _not supposed_ to go messing the argument values. If you're a new user, even seeing the default values is going to be an opening for confusion. This package is an extreme case, for sure.

Regardless, I completely understand the goal of least surprise across the rmarkdown packages. In practice I think it will end up being a pretty small change that's likely to go unnoticed by most users.

**Review by apreshill — APPROVED** [review]
Thanks @gadenbuie! That makes total sense, and I agree it is likely to be undetected by most users.

**Comment by hadley** [comment]
The read is rather long so I haven't read it fully, but I think it's a bad idea to change the standard ordering — even if it does help for these cases, I think users do have a very strong expectation of the ordering based on the hundreds of other docs that they've looked at. If you think long usage is a common problem, I would rather fix that in pkgdown (which will take longer, but it will help everyone, and give a more consistent experience to the user).

**Comment by hadley** [comment]
Also if this is true:

> The problem is definitely exacerbated in gradethis, where for many functions you are not supposed to go messing with the argument values

We should talk about it so we can fix it at the API design level, rather than the documentation level.

**Comment by gadenbuie** [comment]
> If you think long usage is a common problem

I don't think long usage is the problem, rather that the Functions or Methods section, when provided and used well, serve as a better entry point to understanding the purpose of a function than the function signature or a list of function signatures.

That said, I don't think `@describeIn` is a widely used feature. Authors instead tend to find ways to try to fit one-line function descriptions into the `@description` paragraph, e.g. [ymlthis groups distill options](https://ymlthis.r-lib.org/reference/yml_distill_opts.html) but describes functions individually in the description heading. A bullet list — like the one created when using `@describeIn` — would be easier to parse.

After a small survey of pkgdown pages that use this feature, I doubt that a global change would be a net positive improvement, though. `@describeIn` doesn't appear to be used consistently or in a way that will definitely look good across many pages.

> Also if this is true: ... we [should] fix it at the API design level

This is much less true now than it was a few weeks ago, but yes, because gradethis users write code that's evaluated in a non-local environment, the default values and function signature are _not_ a good entry point to understanding the function or its expected usage.

**Comment by hadley** [comment]
> This is much less true now than it was a few weeks ago, but yes, because gradethis users write code that's evaluated in a non-local environment, the default values and function signature are not a good entry point to understanding the function or its expected usage.

IMO, that's a strong signal that the gradethis API could/should be changed rather than its documentation website.

**Comment by gadenbuie** [comment]
Thanks for the feedback, everyone. I still think that it would be generally beneficial for the "Functions" section to appear higher up on the reference page. That said, I understand the need for consistency with R's help pages and other packages.

**Pull request closed without merging**